src/librustc/middle/dead.rs

```diff
@@ -558,8 +558,20 @@ impl<'a, 'tcx> Visitor<'tcx> for DeadVisitor<'a, 'tcx> {
 
     fn visit_struct_field(&mut self, field: &'tcx hir::StructField) {
         if self.should_warn_about_field(&field) {
-            self.warn_dead_code(field.id, field.span,
-                                field.name, "field");
+            let did = self.tcx.hir.get_parent_did(field.id);
+            if if let Some(node_id) = self.tcx.hir.as_local_node_id(did) {
+                match self.tcx.hir.find(node_id) {
+                    Some(hir_map::NodeItem(item)) => match item.node {
+                        Item_::ItemUnion(_, _) => false,
```

Comment:
This is throwing out the baby with the bathwater, union fields can be legitimately unused.

Ideally, use of a union field in a struct expression (`let u = U { a: 10 }`) should mark it as "potentially used", and uses in field accesses/patterns (`u.b`/`let U { b } = u`) should mark the field as "used" and additionally mark all its "potentially used" siblings as "used".